opencog/rule-engine/forwardchainer/ForwardChainer.cc

```diff
@@ -62,81 +67,79 @@ Logger* ForwardChainer::getLogger()
 }
 
 /**
- * Does one step forward chaining
+ * Does one step forward chaining *
  *
- * @param fcb a concrete implementation of of ForwardChainerCallBack class 
+ * @return An unordered sets of result of applying a particular selected rule.
  */
-void ForwardChainer::do_step(ForwardChainerCallBack& fcb)
+UnorderedHandleSet ForwardChainer::do_step(bool search_focus_set/* = false*/)
 {
 
     if (_fcmem.get_cur_source() == Handle::UNDEFINED) {
         _log->info("[ForwardChainer] No current source, step "
                    "forward chaining aborted.");
-        return;
+        return {};
     }
 
     _log->info("[ForwardChainer] Next source %s",
                _fcmem.get_cur_source()->toString().c_str());
 
-    // Choose matching rules whose input matches with the source.
-    vector<Rule*> matched_rules = fcb.choose_rules(_fcmem);
-    _log->info("[ForwardChainer] Found matching rule");
-
-    //! If no rules matches the pattern of the source,
-    //! set all rules for candidacy to be selected by the proceeding step.
-    //! xxx this decision is maded based on recent discussion.I
-    //! it might still face some changes.
-    if (matched_rules.empty()) {
-        _log->info("[ForwardChainer] No matching rule found. "
-                   "Setting all rules as candidates.");
-        matched_rules = _fcmem.get_rules();
-    }
-
-    // Select a rule amongst the matching rules by tournament selection
-    map<Rule*, float> rule_weight;
-    for (Rule* r : matched_rules) {
-        rule_weight[r] = r->get_weight();
-    }
-
     _log->info("[ForwardChainer] Selecting a rule from the set of "
                "candidate rules.");
+
     Rule* r = _rec.tournament_select(rule_weight);
+    _log->info("[ForwardChainer] Selected rule is %s",
+               (r->get_handle())->toShortString().c_str());
+
     _fcmem.set_cur_rule(r);
-    _log->info("[ForwardChainer] Selected rule is %s", (r->get_handle())->toShortString().c_str());
+    //TODO Should we use all derived rules or what?
+    UnorderedHandleSet products;
 
-    //!TODO Find/add premises?
+    HandleSeq derived_rhandles = derive_rules(_fcmem.get_cur_source(), r);
 
-    //! Apply rule.
-    _log->info("[ForwardChainer] Applying chosen rule %s",
-               (r->get_handle())->toShortString().c_str());
-    HandleSeq product = fcb.apply_rule(_fcmem);
+    //try sub-atom unification on failure to full unification
```

Comment:
@misgeatgit Does this mean try unification using sub-atoms of the source? Remember we agreed on the call last week that the full source atom should unify somewhere within the rule premises. Right?